From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jonas Karlman <jonas@kwiboo.se>
Date: Wed, 5 Feb 2025 01:24:07 +0000
Subject: rockchip: rk3506: Enable OP-TEE (SPL_OPTEE_IMAGE) for PSCI

Enable SPL_OPTEE_IMAGE on RK3506 so SPL loads OP-TEE from the FIT and
runs U-Boot proper and the kernel in non-secure mode with a working
PSCI implementation. Without a secure monitor in place, the first PSCI
SMC from the kernel traps to an invalid vector and the board hangs
immediately after "Starting kernel".

OP-TEE runs from the bottom of DRAM, which forces a new memory layout:

  0x00001000  OP-TEE (load and entry address)
  0x00080000  SPL (SPL_TEXT_BASE)
  0x00100000  U-Boot proper (TEXT_BASE)
  0x00200000  SPL FIT load address
  0x03f00000  initial stack pointer
  0x03f80000  SPL BSS

Changes:

- Kconfig: select SPL_OPTEE_IMAGE when SPL_FIT is enabled, imply
  ARM_SMCCC and SYSRESET_PSCI, and add SPL_TEXT_BASE = 0x00080000
  along with the matching TEXT_BASE, init SP, BSS start and
  SPL_LOAD_FIT_ADDRESS defaults keyed on it.
- rk3506-u-boot.dtsi: add the op-tee image to the SPL FIT template
  with load/entry at 0x1000.
- rk3506_common.h: define CFG_SYS_SDRAM_RESERVE as 1 MiB to keep the
  region occupied by OP-TEE out of the memory reported to the kernel.
- sdram.c: honour CFG_SYS_SDRAM_RESERVE in dram_init_banksize() by
  offsetting the start of the first bank, in preference to parsing the
  vendor tos_parameter_t structure.
- rkcommon.c: set the BootROM load address of the second image
  (the SPL) in the v2 header to 0x00080000 for rk3506, instead of the
  0xFFFFFFFF default that would place SPL on top of OP-TEE.

Signed-off-by: Austin Lane <vidplace7@gmail.com>
---
 arch/arm/dts/rk3506-u-boot.dtsi | 11 ++++++++++
 arch/arm/mach-rockchip/Kconfig  |  7 ++++++
 arch/arm/mach-rockchip/sdram.c  |  5 ++++-
 boot/Kconfig                    |  1 +
 include/configs/rk3506_common.h |  1 +
 tools/rkcommon.c                | 10 ++++++++-
 6 files changed, 33 insertions(+), 2 deletions(-)

diff --git a/arch/arm/dts/rk3506-u-boot.dtsi b/arch/arm/dts/rk3506-u-boot.dtsi
index 111111111111..222222222222 100644
--- a/arch/arm/dts/rk3506-u-boot.dtsi
+++ b/arch/arm/dts/rk3506-u-boot.dtsi
@@ -29,6 +29,17 @@
 	bootph-all;
 };
 
+#ifdef CONFIG_SPL_OPTEE_IMAGE
+&fit_template {
+	images {
+		op-tee {
+			load = <0x1000>;
+			entry = <0x1000>;
+		};
+	};
+};
+#endif
+
 &grf_pmu {
 	bootph-all;
 };
diff --git a/arch/arm/mach-rockchip/Kconfig b/arch/arm/mach-rockchip/Kconfig
index 111111111111..222222222222 100644
--- a/arch/arm/mach-rockchip/Kconfig
+++ b/arch/arm/mach-rockchip/Kconfig
@@ -332,6 +332,8 @@ config ROCKCHIP_RK3506
 	select BOARD_LATE_INIT
 	select DM_REGULATOR_FIXED
 	select DM_RESET
+	select SPL_OPTEE_IMAGE if SPL_FIT
+	imply ARM_SMCCC if SPL_OPTEE_IMAGE
 	imply BOOTSTD_FULL
 	imply DM_RNG
 	imply ENV_RELOC_GD_ENV_ADDR
@@ -359,6 +361,7 @@ config ROCKCHIP_RK3506
 	imply SPL_SYSCON
 	imply SYS_ARCH_TIMER
 	imply SYSRESET
+	imply SYSRESET_PSCI if SPL_OPTEE_IMAGE
 	help
 	  The Rockchip RK3506 is a ARM-based SoC with a tri-core Cortex-A7.
 
@@ -803,6 +806,7 @@ if ROCKCHIP_COMMON_STACK_ADDR && SPL_SHARES_INIT_SP_ADDR
 config CUSTOM_SYS_INIT_SP_ADDR
 	default 0x63f00000 if SPL_TEXT_BASE = 0x60000000
 	default 0x43f00000 if SPL_TEXT_BASE = 0x40000000
+	default 0x03f00000 if SPL_TEXT_BASE = 0x00080000
 	default 0x03f00000 if SPL_TEXT_BASE = 0x00000000
 
 config SYS_MALLOC_F_LEN
@@ -824,12 +828,14 @@ config TEXT_BASE
 	default 0x60800000 if SPL_TEXT_BASE = 0x60000000
 	default 0x40800000 if SPL_TEXT_BASE = 0x40000000
 	default 0x00800000 if SPL_TEXT_BASE = 0x00000000
+	default 0x00100000 if SPL_TEXT_BASE = 0x00080000
 
 config SPL_TEXT_BASE
 	default 0x60000000 if ROCKCHIP_RK3036 || ROCKCHIP_RK3066 || \
 			      ROCKCHIP_RK3128 || ROCKCHIP_RK3188 || \
 			      ROCKCHIP_RK322X || ROCKCHIP_RV1108
 	default 0x40000000 if ROCKCHIP_RK3576
+	default 0x00080000 if ROCKCHIP_RK3506
 	default 0x00000000
 
 config SPL_HAS_BSS_LINKER_SECTION
@@ -838,6 +844,7 @@ config SPL_HAS_BSS_LINKER_SECTION
 config SPL_BSS_START_ADDR
 	default 0x63f80000 if SPL_TEXT_BASE = 0x60000000
 	default 0x43f80000 if SPL_TEXT_BASE = 0x40000000
+	default 0x03f80000 if SPL_TEXT_BASE = 0x00080000
 	default 0x03f80000 if SPL_TEXT_BASE = 0x00000000
 
 config SPL_BSS_MAX_SIZE
diff --git a/arch/arm/mach-rockchip/sdram.c b/arch/arm/mach-rockchip/sdram.c
index 111111111111..222222222222 100644
--- a/arch/arm/mach-rockchip/sdram.c
+++ b/arch/arm/mach-rockchip/sdram.c
@@ -320,7 +320,7 @@ int dram_init_banksize(void)
 		gd->bd->bi_dram[0].size = ram_top - gd->bd->bi_dram[0].start;
 	}
 #else
-#ifdef CONFIG_SPL_OPTEE_IMAGE
+#if defined(CONFIG_SPL_OPTEE_IMAGE) && !defined(CFG_SYS_SDRAM_RESERVE)
 	struct tos_parameter_t *tos_parameter;
 
 	tos_parameter = (struct tos_parameter_t *)(CFG_SYS_SDRAM_BASE +
@@ -343,6 +343,9 @@ int dram_init_banksize(void)
 	}
 #else
 	gd->bd->bi_dram[0].start = CFG_SYS_SDRAM_BASE;
+#ifdef CFG_SYS_SDRAM_RESERVE
+	gd->bd->bi_dram[0].start += CFG_SYS_SDRAM_RESERVE;
+#endif
 	gd->bd->bi_dram[0].size = top - gd->bd->bi_dram[0].start;
 #endif
 #endif
diff --git a/boot/Kconfig b/boot/Kconfig
index 111111111111..222222222222 100644
--- a/boot/Kconfig
+++ b/boot/Kconfig
@@ -263,6 +263,7 @@ config SPL_LOAD_FIT_ADDRESS
 	default 0x44000000 if ARCH_IMX8M
 	default 0x60080000 if ARCH_ROCKCHIP && SPL_TEXT_BASE = 0x60000000
 	default 0x40200000 if ARCH_ROCKCHIP && SPL_TEXT_BASE = 0x40000000
+	default 0x00200000 if ARCH_ROCKCHIP && SPL_TEXT_BASE = 0x00080000
 	default 0x00200000 if ARCH_ROCKCHIP && SPL_TEXT_BASE = 0x00000000
 	help
 	  Specify the load address of the fit image that will be loaded
diff --git a/include/configs/rk3506_common.h b/include/configs/rk3506_common.h
index 111111111111..222222222222 100644
--- a/include/configs/rk3506_common.h
+++ b/include/configs/rk3506_common.h
@@ -11,6 +11,7 @@
 #define CFG_IRAM_BASE			0xfff80000
 
 #define CFG_SYS_SDRAM_BASE		0
+#define CFG_SYS_SDRAM_RESERVE		SZ_1M
 #define SDRAM_MAX_SIZE			0xc0000000
 
 #ifndef ROCKCHIP_DEVICE_SETTINGS
diff --git a/tools/rkcommon.c b/tools/rkcommon.c
index 111111111111..222222222222 100644
--- a/tools/rkcommon.c
+++ b/tools/rkcommon.c
@@ -362,7 +362,15 @@ static void rkcommon_set_header0_v2(void *buf, struct image_tool_params *params)
 		image_sector_count = image_size_array[i] / RK_BLK_SIZE;
 		hdr->images[i].size_and_off = cpu_to_le32((image_sector_count
 							<< 16) + sector_offset);
-		hdr->images[i].address = 0xFFFFFFFF;
+		/*
+		 * RK3506 loads OP-TEE at the bottom of DRAM, so the SPL
+		 * (second image) must be loaded above it instead of at the
+		 * BootROM default address.
+		 */
+		if (i == 1 && !strcmp(params->imagename, "rk3506"))
+			hdr->images[i].address = cpu_to_le32(0x00080000);
+		else
+			hdr->images[i].address = 0xFFFFFFFF;
 		hdr->images[i].counter = cpu_to_le32(i + 1);
 		image_ptr = buf + sector_offset * RK_BLK_SIZE;
 		do_sha256_hash(image_ptr, image_size_array[i],
-- 
Armbian

